### ISSUE TITLE
Download multiple files at once

### ISSUE THREAD

**Issue by nyuszika7h** [megous/megatools#299]
It would be nice if there was an option to specify how many of the files specified on the command line should be downloaded in parallel. This could be used to work around #286 (if each individual file is under 5 GB) or simply speed up downloads. I can currently do this with `xargs megadl -n1 -P0 <<EOF`, but the output is practically unreadable as the `megadl` processes fight for stdout.

**Comment by megous** [comment]
I'm curious. How do you think would it workaround the limit? Wouldn't the server simply drop the connections anyway?

**Comment by nyuszika7h** [comment]
I would expect so, but no, it doesn't, as long as you start them at the same time. This is a trick people have used with MegaDownloader that I read in a reddit thread.

**Comment by thedarkfalcon** [comment]
I think this would be a great feature as megatools only downloads each file with a single thread. If we can't increase download threads per file, scaling out how many files are downloaded at once would be the next best thing. Currently when I download files on my NAS, I use aria2 with connections per file set to 16 (the max aria2 allows), and 8 parallel downloads. This is mainly due to my geographic location, using a single download stream is garbage :[

I see talk of adding download multithreading here
https://github.com/megous/megatools/issues/188

But both methods of increasing throughput would be good for me :D

**Comment by megous** [comment]
@thedarkfalcon So have you tried xargs -P8 -n1 trick for example. Does it help with bandwidth with mega?

**Comment by thedarkfalcon** [comment]
@megous yes it does, but as @nyuszika7h mentioned, though stdout is impossible to make sense of. But my main annoyance with that method is, it doesn't help if I want to just get one file quickly! D:

I have done testing on my internet connection with aria2, normally per download thread I get about 200kBytes/s, so with aria2 running 16 threads per file I get about 3.2mBytes/s. If I continue scaling, 8 - 10 parallel downloads (each with 16 threads) seems to max out the speed I can achieve (around 32mBytes/s).
tl;dr I get 200kBytes/s with a single stream download, but up to 32mBytes/s if I scale out massively.

**Comment by megous** [comment]
OTOH, I'm not sure how common this is (so far only a round 3-4 people asked for this and only you presented some numbers :)), so while this will complicate the code quite a bit, it may not benefit that many people.

Seems like some kind of weird per connection QoS in action. Have you tried asking your ISP to fix it's service? You can workaround the limit anyway, so they may as well implement something that's not per connection, but per link, like most other ISPs do.

**Comment by thedarkfalcon** [comment]
It's purely geographic (I live in New Zealand), and isn't related to available bandwidth, but rather Round Trip Time. In a straight line; the distance to USA  is 12,000km, and to Europe nearly 20,000km (Mega servers are hosted in Germany). NZ has no direct routes pretty much anywhere except to Australia, so even though the USA is 12,000km east, data starts going west first. RTT to Europe is about 600ms so you can imagine why a single TCP stream waiting for an ACK is extremely slow. Another way to increase throughput would be if the TCP MTU were to ever be increased - so larger chunks of data come through at a time, but I doubt that will happen any time soon. Hence the only way to pump the data through faster is to really scale out the number of streams downloading.

So ironically even though Mega has an NZ domain and is run out of NZ, it has basically the worst RTT possible to it.

**Comment by megous** [comment]
Very interesting. Does tuning the TCP settings help? Like increasing default/max receive window size and enabling SACK. Megatools will just pull data from the remote side, with no other communication aside from rare TLS re-keying. With 32MB window you should get 32MB/s even with 600ms RTT.

**Comment by thedarkfalcon** [comment]
To be honest I have never really tried hacking around with TCP options that much except trying to force the QoS bit.

**Comment by megous** [comment]
@thedarkfalcon Did you tinker with it? I'm curious how it went, if these things really help real world high latency links across the globe, if you don't mind sharing.

**Comment by megous** [comment]
Parallel upload is now implemented. Parallel download should now be easy to do, too.

**Comment by thedarkfalcon** [comment]
@megous that's good news. No sorry I haven't had a chance to tinker with it.
